templates/base/client/.rtag/Forms.tsx.hbs

```diff
@@ -0,0 +1,296 @@
+import React, { useState, useContext } from "react";
+import { toast } from "react-toastify";
+import { RtagContext } from "./context";
+import { Response } from "./base";
+import {
+  {{#each methods}}
+  {{makeRequestName @key}},
+  {{/each}}
+  {{#each types}}
+  {{@key}},
+  {{/each}}
+} from "./types";
+
+type MethodProps<T> = {
+  method: string;
+  submit: (value: T) => Promise<Response>;
+  initialize: () => T;
+  children: (value: T, update: (value: T) => void) => JSX.Element;
+};
+type BaseProps<T> = {
+  value: T;
+  update: (value: T) => void;
+};
+type ContainerProps<T> = {
+  initialVal: T;
+  children: (value: T, update: (value: T) => void) => JSX.Element;
+};
+type ArrayProps<T> = BaseProps<T[]> & ContainerProps<T>;
+type OptionalProps<T> = BaseProps<T | undefined> & ContainerProps<T>;
+type EnumProps = BaseProps<number> & { enum: object };
+
+export function MethodForm<T>(props: MethodProps<T>) {
+  const [value, setValue] = useState<T>(props.initialize());
+  console.log("MethodForm", value);
+  return (
+    <div className="method-form">
+      <h3>{props.method}</h3>
+      {props.children(value, setValue)}
+      <button
+        onClick={async () => {
+          const res = await props.submit(value);
+          if (res.type === "error") {
+            toast.error(res.error);
+          }
+          setValue(props.initialize());
+        }}
+      >
+        Submit
+      </button>
+    </div>
+  );
+}
+
+function ArrayInput<T>(props: ArrayProps<T>) {
+  function swapArgs(i1: number, i2: number) {
+    [props.value[i1], props.value[i2]] = [props.value[i2], props.value[i1]];
+    props.update(props.value);
+  }
+  function deleteArg(i: number) {
+    props.value.splice(i, 1);
+    props.update(props.value);
+  }
+  return (
+    <div className="array-input">
+      {props.value.map((item, i) => (
+        <div key={i} className="array-item">
+          {props.children(item, (val) => {
+            props.update(Object.assign(props.value, { [i]: val }));
+          })}
+          <button disabled={i === 0} onClick={() => swapArgs(i, i - 1)}>
+            &#8593;
+          </button>
+          <button disabled={i === props.value.length - 1} onClick={() => swapArgs(i, i + 1)}>
```

Comment:
convention is to destructure props at the top of each component like `const { value, update } = props`. i think it helps readability--e.g. my initial thought when seeing `props.update` is that you were calling _update_ on the props itself for some reason. i know that's not what you're doing but i basically never see props.anything so i was confused for a second